src/compile/data/parse.ts

```diff
@@ -179,129 +157,88 @@ Description of the dataflow (http://asciiflow.com/):
 */
 
 export function parseData(model: Model): DataComponent {
-  const root = parseRoot(model, model.component.data.sources);
+  let head = parseRoot(model, model.component.data.sources);
 
   const outputNodes = model.component.data.outputNodes;
   const outputNodeRefCounts = model.component.data.outputNodeRefCounts;
 
-  // the current head of the tree that we are appending to
-  let head = root;
-
   // Default discrete selections require an identifier transform to
   // uniquely identify data points as the _id field is volatile. Add
   // this transform at the head of our pipeline such that the identifier
   // field is available for all subsequent datasets. Additional identifier
   // transforms will be necessary when new tuples are constructed
   // (e.g., post-aggregation).
   if (requiresSelectionId(model) && !model.parent) {
-    const ident = new IdentifierNode();
-    ident.parent = head;
-    head = ident;
+    head = new IdentifierNode(head);
   }
 
   // HACK: This is equivalent for merging bin extent for union scale.
   // FIXME(https://github.com/vega/vega-lite/issues/2270): Correctly merge extent / bin node for shared bin scale
   const parentIsLayer = model.parent && isLayerModel(model.parent);
   if (isUnitModel(model) || isFacetModel(model)) {
     if (parentIsLayer) {
-      const bin = BinNode.makeBinFromEncoding(model);
-      if (bin) {
-        bin.parent = head;
-        head = bin;
-      }
+      head = BinNode.makeBinFromEncoding(head, model) || head;
     }
   }
 
   if (model.transforms.length > 0) {
-    const {first, last} = parseTransformArray(model);
-    first.parent = head;
-    head = last;
+    head = parseTransformArray(head, model);
   }
 
-  const parse = ParseNode.make(model);
+  const parse = ParseNode.make(head, model);
   if (parse) {
-    parse.parent = head;
     head = parse;
   }
 
   if (isUnitModel(model)) {
-    for (const geojson of GeoJSONNode.makeAll(model)) {
-      geojson.parent = head;
-      head = geojson;
-    }
-
-    for (const geopoint of GeoPointNode.makeAll(model)) {
-      geopoint.parent = head;
-      head = geopoint;
-    }
+    head = GeoJSONNode.makeAll(head, model);
```

Comment:
makeAll gives a feel like factory pattern, but I think it's better named `parseAll()` ? (More consistent with `parseTransformArray` above.)

Reply:
Done